scalacheck/src/main/scala/karat/scalacheck/Scalacheck.scala

```diff
@@ -41,4 +43,39 @@ object Scalacheck {
     )
     if (problem == null) Prop.passed else Prop(problem.getError)
   }
-}
\ No newline at end of file
+
+  def checkFormula[F[_] : Monad, Action, State, Response](actions: List[Action], initial: F[State], step: (Action, State) => F[Step[State, Response]])(
+    formula: Formula[Info[Action, State, Response]]
+  ): F[Prop] =
+    for {
+      current <- initial
+      result <- checkFormula(new ScalacheckStepResultManager[Info[Action, State, Response]](), step, actions, current, formula)
+    } yield result
+
+  def checkFormula[F[_] : Monad, Action, State, Response](
+    resultManager: ScalacheckStepResultManager[Info[Action, State, Response]],
+    step: (Action, State) => F[Step[State, Response]],
+    actions: List[Action],
+    current: State,
+    formula: Formula[Info[Action, State, Response]]
+  ): F[Prop] = Monad[F].tailRecM((actions, current, formula)) { case (actions, current, formula) =>
+    actions match {
+      case Nil =>
+        resultToProp(CheckKt.leftToProve(resultManager, formula)).asRight.pure
+      case action :: rest => step(action, current).flatMap {
```

Comment:
We can directly use the `case` in the `tailRecM`:

```scala
Monad[F].tailRecM((actions, current, formula)) { 
  case (Nil, _, formula) =>
    resultToProp(CheckKt.leftToProve(resultManager, formula)).asRight.pure
  case (action :: rest, current, formula) =>
    step(action, current).flatMap {
//...
```

Reply:
TIL